File: native-views/react-native-image/android/src/main/java/com/margelo/nitro/onekeyimage/OneKeyImageMemoryVariants.kt
package com.margelo.nitro.onekeyimage

import kotlin.math.max
import kotlin.math.min

internal data class OneKeyImageMemoryFamilyKey(
  val rawUrl: String,
  val headersDigest: String?,
) {
  companion object {
    fun from(rawUrl: String, headersJson: String?): OneKeyImageMemoryFamilyKey =
      OneKeyImageMemoryFamilyKey(
        rawUrl = rawUrl,
        headersDigest = OneKeyImageModel.remoteHeadersDigest(
          OneKeyImageModel.headers(headersJson),
        ),
      )
  }
}

internal data class OneKeyImageMemoryVariant(
  val requestUrl: String,
  val width: Int,
  val height: Int,
  val round: Boolean,
) {
  val area: Long
    get() = width.toLong() * height.toLong()
}

/**
 * A bounded index of size-specific Glide resources that may still be in memory.
 *
 * Glide remains the source of truth. Entries are only hints used to reproduce a
 * candidate EngineKey; a synchronous cache miss removes the stale hint.
 */
internal object OneKeyImageMemoryVariantRegistry {
  private const val MAX_FAMILIES = 1024
  private const val MAX_VARIANTS_PER_FAMILY = 8
  private const val MAX_ASPECT_RATIO_DRIFT = 1.1

  private val variantsByFamily =
    LinkedHashMap<OneKeyImageMemoryFamilyKey, MutableList<OneKeyImageMemoryVariant>>(
      16,
      0.75f,
      true,
    )

  @Synchronized
  fun record(family: OneKeyImageMemoryFamilyKey, variant: OneKeyImageMemoryVariant) {
    if (variant.width <= 0 || variant.height <= 0) return
    val variants = variantsByFamily.getOrPut(family) { mutableListOf() }
    variants.remove(variant)
    variants.add(variant)
    while (variants.size > MAX_VARIANTS_PER_FAMILY) variants.removeAt(0)
    while (variantsByFamily.size > MAX_FAMILIES) {
      val eldest = variantsByFamily.entries.firstOrNull()?.key ?: break
      variantsByFamily.remove(eldest)
    }
  }

  @Synchronized
  fun remove(family: OneKeyImageMemoryFamilyKey, variant: OneKeyImageMemoryVariant) {
    val variants = variantsByFamily[family] ?: return
    variants.remove(variant)
    if (variants.isEmpty()) variantsByFamily.remove(family)
  }

  @Synchronized
  fun candidates(
    family: OneKeyImageMemoryFamilyKey,
    target: OneKeyImageMemoryVariant,
  ): List<OneKeyImageMemoryVariant> = orderedCandidates(
    variants = variantsByFamily[family].orEmpty(),
    target = target,
  )

  @Synchronized
  fun clear() {
    variantsByFamily.clear()
  }

  internal fun orderedCandidates(
    variants: List<OneKeyImageMemoryVariant>,
    target: OneKeyImageMemoryVariant,
  ): List<OneKeyImageMemoryVariant> {
    val compatible = variants
      .asSequence()
      .filter { it != target }
      .filter { candidate -> target.round || !candidate.round }
      .filter { candidate -> aspectRatioIsCompatible(candidate, target) }
      .distinct()
      .toList()

    val larger = compatible
      .filter { it.width >= target.width && it.height >= target.height }
      .minWithOrNull(
        compareBy<OneKeyImageMemoryVariant> { it.area }
          .thenBy { if (it.round == target.round) 0 else 1 },
      )
    val smaller = compatible
      .filter { it.width <= target.width && it.height <= target.height }
      .maxWithOrNull(
        compareBy<OneKeyImageMemoryVariant> { it.area }
          .thenByDescending { if (it.round == target.round) 0 else 1 },
      )
    return listOfNotNull(larger, smaller).distinct()
  }

  private fun aspectRatioIsCompatible(
    candidate: OneKeyImageMemoryVariant,
    target: OneKeyImageMemoryVariant,
  ): Boolean {
    val candidateRatio = candidate.width.toDouble() / candidate.height.toDouble()
    val targetRatio = target.width.toDouble() / target.height.toDouble()
    return max(candidateRatio, targetRatio) / min(candidateRatio, targetRatio) <=
      MAX_ASPECT_RATIO_DRIFT
  }
}

File: native-views/react-native-image/ios/OneKeyImageMemoryVariants.swift
import Foundation

struct OneKeyImageMemoryFamilyKey: Hashable, Sendable {
  let rawURL: URL
  let headersDigest: String?

  init(rawURL: URL, headersJson: String?) {
    self.rawURL = rawURL
    headersDigest =
      OneKeyImageHeadersIdentity(
        headers: OneKeyImageRequestContext.headers(from: headersJson)
      ).cacheKeyComponent
  }
}

struct OneKeyImageMemoryVariant: Hashable, Sendable {
  let requestURL: URL
  let pixelWidth: Int
  let pixelHeight: Int

  init(requestURL: URL, thumbnailPixelSize: CGSize) {
    self.requestURL = requestURL
    pixelWidth = max(Int(thumbnailPixelSize.width.rounded()), 1)
    pixelHeight = max(Int(thumbnailPixelSize.height.rounded()), 1)
  }

  var thumbnailPixelSize: CGSize {
    CGSize(width: pixelWidth, height: pixelHeight)
  }

  var area: Int64 { Int64(pixelWidth) * Int64(pixelHeight) }
}

/// Bounded hints for reproducing size-specific SDWebImage memory-cache keys.
/// SDImageCache remains authoritative; stale hints are removed on probe miss.
final class OneKeyImageMemoryVariantRegistry: @unchecked Sendable {
  static let shared = OneKeyImageMemoryVariantRegistry()

  private static let maximumFamilies = 1024
  private static let maximumVariantsPerFamily = 8
  private static let maximumAspectRatioDrift = 1.1

  private let lock = NSLock()
  private var variantsByFamily: [OneKeyImageMemoryFamilyKey: [OneKeyImageMemoryVariant]] = [:]
  private var familyAccessOrder: [OneKeyImageMemoryFamilyKey] = []

  func record(_ variant: OneKeyImageMemoryVariant, for family: OneKeyImageMemoryFamilyKey) {
    lock.lock()
    defer { lock.unlock() }
    var variants = variantsByFamily[family] ?? []
    variants.removeAll { $0 == variant }
    variants.append(variant)
    if variants.count > Self.maximumVariantsPerFamily {
      variants.removeFirst(variants.count - Self.maximumVariantsPerFamily)
    }
    variantsByFamily[family] = variants
    touch(family)
    while familyAccessOrder.count > Self.maximumFamilies {
      let oldest = familyAccessOrder.removeFirst()
      variantsByFamily.removeValue(forKey: oldest)
    }
  }

  func remove(_ variant: OneKeyImageMemoryVariant, for family: OneKeyImageMemoryFamilyKey) {
    lock.lock()
    defer { lock.unlock() }
    guard var variants = variantsByFamily[family] else { return }
    variants.removeAll { $0 == variant }
    if variants.isEmpty {
      variantsByFamily.removeValue(forKey: family)
      familyAccessOrder.removeAll { $0 == family }
    } else {
      variantsByFamily[family] = variants
    }
  }

  func candidates(
    for family: OneKeyImageMemoryFamilyKey,
    target: OneKeyImageMemoryVariant,
    excluding exactVariants: Set<OneKeyImageMemoryVariant>
  ) -> [OneKeyImageMemoryVariant] {
    lock.lock()
    let variants = variantsByFamily[family] ?? []
    if !variants.isEmpty { touch(family) }
    lock.unlock()
    return Self.orderedCandidates(
      variants: variants,
      target: target,
      excluding: exactVariants
    )
  }

  func clear() {
    lock.lock()
    variantsByFamily.removeAll()
    familyAccessOrder.removeAll()
    lock.unlock()
  }

  static func orderedCandidates(
    variants: [OneKeyImageMemoryVariant],
    target: OneKeyImageMemoryVariant,
    excluding exactVariants: Set<OneKeyImageMemoryVariant> = []
  ) -> [OneKeyImageMemoryVariant] {
    var seen: Set<OneKeyImageMemoryVariant> = []
    let compatible = variants.reversed().filter {
      seen.insert($0).inserted && !exactVariants.contains($0) && aspectRatioIsCompatible($0, target)
    }
    let larger =
      compatible
      .filter { $0.pixelWidth >= target.pixelWidth && $0.pixelHeight >= target.pixelHeight }
      .min { lhs, rhs in lhs.area < rhs.area }
    let smaller =
      compatible
      .filter { $0.pixelWidth <= target.pixelWidth && $0.pixelHeight <= target.pixelHeight }
      .max { lhs, rhs in lhs.area < rhs.area }
    var result: [OneKeyImageMemoryVariant] = []
    if let larger { result.append(larger) }
    if let smaller, smaller != larger { result.append(smaller) }
    return result
  }

  private static func aspectRatioIsCompatible(
    _ candidate: OneKeyImageMemoryVariant,
    _ target: OneKeyImageMemoryVariant
  ) -> Bool {
    let candidateRatio = Double(candidate.pixelWidth) / Double(candidate.pixelHeight)
    let targetRatio = Double(target.pixelWidth) / Double(target.pixelHeight)
    return max(candidateRatio, targetRatio) / min(candidateRatio, targetRatio)
      <= maximumAspectRatioDrift
  }

  private func touch(_ family: OneKeyImageMemoryFamilyKey) {
    familyAccessOrder.removeAll { $0 == family }
    familyAccessOrder.append(family)
  }
}
